jena-text/src/main/java/org/apache/jena/query/text/TextIndexES.java

```diff
@@ -0,0 +1,394 @@
+        Map<String, Node> result = new HashMap<>();
+
+        if(uri != null) {
+            response = client.prepareGet(indexName, docDef.getEntityField(), uri).get();
+            if(response != null && !response.isSourceEmpty()) {
+                String entityField = response.getId();
+                Node entity = NodeFactory.createURI(entityField) ;
+                result.put(docDef.getEntityField(), entity);
+                Map<String, Object> source = response.getSource();
+                for (String field: docDef.fields()) {
+                    Object fieldResponse = source.get(field);
+
+                    if(fieldResponse == null) {
+                        //We wont return it.
+                        continue;
+                    }
+                    else if(fieldResponse instanceof List<?>) {
+                        //We are storing the values of fields as a List always.
+                        //If there are values stored in the list, then we return the first value,
+                        // else we do not include the field in the returned Map of Field -> Node Mapping
+                        List<?> responseList = (List<?>)fieldResponse;
+                        if(responseList != null && responseList.size() > 0) {
+                            String fieldValue = (String)responseList.get(0);
+                            Node fieldNode = NodeFactoryExtra.createLiteralNode(fieldValue, null, null);
+                            result.put(field, fieldNode);
+                        }
+                    }
+                }
+            }
+        }
+
+        return result;
+    }
+
+    @Override
+    public List<TextHit> query(Node property, String qs) {
+
+        return query(property, qs, MAX_RESULTS);
+    }
+
+    /**
+     * Query the ElasticSearch for the given Node, with the given query String and limit.
+     * @param property the node property to make a search for
+     * @param qs the query string
+     * @param limit limit on the number of records to return
+     * @return List of {@link TextHit}s containing the documents that have been found
+     */
+    @Override
+    public List<TextHit> query(Node property, String qs, int limit) {
+
+        qs = parse(qs);
+        LOGGER.debug("Querying ElasticSearch for QueryString: " + qs);
+        SearchResponse response = client.prepareSearch(indexName)
+                .setTypes(docDef.getEntityField())
+                .setQuery(QueryBuilders.queryStringQuery(qs))
+                .setFrom(0).setSize(limit)
+                .get();
+
+        List<TextHit> results = new ArrayList<>() ;
+        for (SearchHit hit : response.getHits()) {
+
+            Node literal;
+            String field = (property != null) ? docDef.getField(property) : docDef.getPrimaryField();
+            List<String> value = (List<String>)hit.getSource().get(field);
+            if(value != null) {
+                literal = NodeFactory.createLiteral(value.get(0));
```

Comment:
So @osma. I need to understand a bit more clearly the importance of the returned Literal value here. Whatever I could debug, I found that this literal value is actually not used. The only thing, IMO, that matters is the Subject URI. I am sure I am missing something, but cant seem to point exactly what. So can you please throw some more light on the importance of returning a specific literal value. In my case, when I do a search like this: (rdfs:label 'X1'), I will get back the label specific field that is actually an Array of values (because I utilize a single document to index multiple values). I can search through this arraylist to find ALL the matched literal values and create a TextIndex instance for each matched Literal, but I dont want to do it if it is not necessary. Please let me know your thoughts around it.